include/boost/math/special_functions/prime_sieve.hpp

```diff
@@ -0,0 +1,402 @@
+// Copyright 2020 Matt Borland
+//
+// Use, modification and distribution are subject to the
+// Boost Software License, Version 1.0.
+// (See accompanying file LICENSE_1_0.txt
+// or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#ifndef BOOST_MATH_SPECIAL_FUNCTIONS_PRIME_SIEVE_HPP
+#define BOOST_MATH_SPECIAL_FUNCTIONS_PRIME_SIEVE_HPP
+
+#include <boost/math/special_functions/prime.hpp>
+#include <boost/assert.hpp>
+#include <vector>
+#include <iterator>
+#include <cmath>
+#include <thread>
+#include <memory>
+#include <future>
+
+#ifdef _MSVC_LANG
+#if _MSVC_LANG >= 201703 // _MSVC_LANG == __cplusplus: https://devblogs.microsoft.com/cppblog/msvc-now-correctly-reports-__cplusplus/
+#include <execution>
+#endif
+#else
+#if __cplusplus >= 201703
+#include <execution>
+#endif
+#endif
+
+namespace boost { namespace math { namespace detail
+{
+// https://mathworld.wolfram.com/SieveofEratosthenes.html
+// https://www.cs.utexas.edu/users/misra/scannedPdf.dir/linearSieve.pdf
+template<class Z, class Container>
+void linear_sieve(Z upper_bound, Container &c)
+{
+    size_t least_divisors_size{static_cast<size_t>(upper_bound + 1)};
+    std::unique_ptr<Z[]> least_divisors{new Z[least_divisors_size]{0}};
+
+    for (Z i{2}; i <= upper_bound; ++i)
+    {
+        if (least_divisors[i] == 0)
+        {
+            least_divisors[i] = i;
+            c.emplace_back(i);
+        }
+
+        for (size_t j{}; j < least_divisors_size; ++j)
+        {
+            if (j >= c.size())
+            {
+                break;
+            }
+
+            else if (c[j] > least_divisors[i])
+            {
+                break;
+            }
+
+            else if (i * c[j] > upper_bound)
+            {
+                break;
+            }
+
+            else
+            {
+                least_divisors[i * c[j]] = c[j];
+            }
+        }
+    }
+}
+
+//https://core.ac.uk/download/pdf/62440589.pdf
+template<class Z, class PrimeContainer, class Container>
+void mask_sieve(Z lower_bound, Z upper_bound, const PrimeContainer& primes, Container &c)
+{
+    Z limit {static_cast<Z>(std::floor(std::sqrt(static_cast<double>(upper_bound)))) + 1};
+
+    size_t primes_size {};
+    auto it{primes.begin()};
+    while(it != primes.end() && *it < limit)
+    {
+        ++primes_size;
+        ++it;
+    }
+
+    const size_t n {static_cast<size_t>(upper_bound - lower_bound + 1)};
+    std::unique_ptr<bool[]> is_prime {new bool[n]};
+    memset(is_prime.get(), true, sizeof(*is_prime.get()) * (n));
```

Comment:
Using `vector` would simplify this code, too. These three lines would become:

`std::vector<bool> is_prime(upper_bound - lower_bound + 1, true);`

Reply:
It would also get the benefits and quirks of `std::vector<bool>`.

Reply:
I remember that Herb Sutter used to recommend against using `std::vector<bool>` because it was standard non-conforming. I found this current implementation to have slightly lower memory usage and complexity than his recommended `std::vector<char>`. The difference was again in micro-optimization territory.

Reply:
Yeah, right, fair point.